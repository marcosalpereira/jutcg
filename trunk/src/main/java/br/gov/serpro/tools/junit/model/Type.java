package br.gov.serpro.tools.junit.model;

public class Type {
	private String name;

	private String variableName;

	private boolean primitive;

	private String fullName;

	public String getNewValue() {
		if (!isPrimitive()) {
			if (isCollection()) {
				return String.format("new %s()", getCollectionImpl());
			}
			return String.format("new %s(1)", getName());
		}

		if (getName().equals("boolean")) {
			return "false";
		}

		return "0";
	}

	private String getCollectionImpl() {
		if (isSet()) return "Hash" + name;
		if (isList()) return "Array" + name;
		return name;
	}

	private boolean isCollection() {
		return isList() || isSet();
	}

	private boolean isSet() {
		return name.startsWith("Set");
	}

	private boolean isList() {
		return name.startsWith("List");
		}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
		this.variableName = defineVariableName();
	}

	public void setPrimitive(boolean primitive) {
		this.primitive = primitive;
	}

	public boolean isPrimitive() {
		return primitive;
	}

	public String getVariableName() {
		return variableName;
	}

	private String defineVariableName() {

		if (name == null) {
			return null;
		} else {
			return name.substring(0, 1).toLowerCase() + name.substring(1);
		}
	}

	/** {@inheritDoc} */
	@Override
	public String toString() {
		return name;
	}

	public void setFullName(String fullName) {
		this.fullName = fullName;
	}

	public String getFullName() {
		return fullName;
	}



}
