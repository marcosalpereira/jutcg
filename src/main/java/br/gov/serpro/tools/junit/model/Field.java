package br.gov.serpro.tools.junit.model;

import java.util.ArrayList;
import java.util.List;

import br.gov.serpro.tools.junit.util.GeneratorHelper;


public class Field implements Comparable<Field>{
	/**
	 * Field name.
	 */
	private String name;

	/**
	 * Field type.
	 */
	private Type type;

	/**
	 * Is field static
	 */
	private boolean ztatic;

	/**
	 * Field's annotations.
	 */
	private List<String> annotations = new ArrayList<String>();

	/**
	 * Field protection.
	 */
	private Protection protection;

	/**
	 * Assigned value, if known.
	 */
	private String writtenValue;

	public void setProtection(Protection protection) {
		this.protection = protection;
	}

	public Protection getProtection() {
		return protection;
	}

	public List<String> getAnnotations() {
		return annotations;
	}

	public void setAnnotations(List<String> annotations) {
		this.annotations = annotations;
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public Type getType() {
		return type;
	}

	public void setType(Type type) {
		this.type = type;
	}

	public String variableNameForType() {
		if (getType() == null) return null;
		return getType().getVariableName();
	}

	public boolean isStatic() {
		return ztatic;
	}

	public void setStatic(boolean ztatic) {
		this.ztatic = ztatic;
	}

	public boolean isPrivate() {
		return protection == Protection.PRIVATE;
	}

	public String getWrittenValue() {
		return writtenValue;
	}

	public void setWrittenValue(String writtenValue) {
		this.writtenValue = writtenValue;
	}

	public boolean isWrittenValueKnown() {
		return writtenValue != null;
	}

	public boolean isWrittenValueNullLiteral() {
		return getWrittenValue().equals("null");
	}

	public boolean isWrittenValueBooleanLiteral() {
		return isWrittenValueTrueLiteral()
			|| isWrittenValueFalseLiteral();
	}

	private boolean isWrittenValueFalseLiteral() {
		return getWrittenValue().equals("false");
	}

	public boolean isWrittenValueTrueLiteral() {
		return getWrittenValue().equals("true");
	}

	@Override
	public String toString() {
		return this.name;
	}

	public String getSetter() {
		return "set" + GeneratorHelper.upperCaseFirstChar(name);
	}

	public String getGetter() {
		if ("boolean".equals(this.getType().getName())) {
			return "is" + GeneratorHelper.upperCaseFirstChar(name);
		}
		return "get" + GeneratorHelper.upperCaseFirstChar(name);
	}

    @Override
    public int compareTo(Field o) {
        return getName().compareTo(o.getName());
    }
}
