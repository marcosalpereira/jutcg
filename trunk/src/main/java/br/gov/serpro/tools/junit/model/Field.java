package br.gov.serpro.tools.junit.model;

import java.util.ArrayList;
import java.util.List;

import br.gov.serpro.tools.junit.GeneratorHelper;


public class Field {
	/**
	 * Field name.
	 */
	private String name;

	/**
	 * Field type.
	 */
	private Type type;

	/**
	 * Is field static
	 */
	private boolean ztatic;

	/**
	 * Field's annotations.
	 */
	private List<String> annotations = new ArrayList<String>();

	/**
	 * Field protection.
	 */
	private Protection protection;

	public void setProtection(Protection protection) {
		this.protection = protection;
	}

	public Protection getProtection() {
		return protection;
	}

	public List<String> getAnnotations() {
		return annotations;
	}

	public void setAnnotations(List<String> annotations) {
		this.annotations = annotations;
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public Type getType() {
		return type;
	}

	public void setType(Type type) {
		this.type = type;
	}

	public String variableNameForType() {
		if (getType() == null) return null;
		return getType().getVariableName();
	}

	public boolean isStatic() {
		return ztatic;
	}

	public void setStatic(boolean ztatic) {
		this.ztatic = ztatic;
	}

	public boolean isPrivate() {
		return protection == Protection.PRIVATE;
	}
	
	@Override
	public String toString() {
		return this.name;
	}

	public String getSetter() {
		return "set" + GeneratorHelper.upperCaseFirstChar(name);
	}

	public String getGetter() {
		if ("boolean".equals(this.getType().getName())) {
			return "is" + GeneratorHelper.upperCaseFirstChar(name);
		}
		return "get" + GeneratorHelper.upperCaseFirstChar(name);
	}
}
