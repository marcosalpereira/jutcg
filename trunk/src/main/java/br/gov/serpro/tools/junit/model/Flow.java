package br.gov.serpro.tools.junit.model;

import java.util.ArrayList;
import java.util.List;
import java.util.SortedSet;
import java.util.TreeSet;

public class Flow {

	/**
	 * Flow name.
	 */
	private String name = "";

	/**
	 * Invocations for this flow.
	 */
	private List<FieldMethodInvocation> invocations = new ArrayList<FieldMethodInvocation>();

	/**
	 * Method that this flow belongs to.
	 */
	private Method method;

	/**
	 * Fields read on this flow.
	 */
	private SortedSet<Field> readFields = new TreeSet<Field>();

	/**
	 * Fields written on this flow.
	 */
	private SortedSet<Field> writtenFields = new TreeSet<Field>();


	public SortedSet<Field> getWrittenFields() {
		return writtenFields;
	}

	public void setWrittenFields(SortedSet<Field> writtenFields) {
		this.writtenFields = writtenFields;
	}

	public SortedSet<Field> getReadFields() {
		return readFields;
	}

	public void setReadFields(SortedSet<Field> usedFields) {
		this.readFields = usedFields;
	}

	public List<FieldMethodInvocation> getInvocations() {
		return invocations;
	}

	public void setInvocations(List<FieldMethodInvocation> invocations) {
		this.invocations = invocations;
	}

	public Method getMethod() {
		return this.method;
	}

	public void setMethod(Method method) {
		this.method = method;
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	/**
	 * Retorna o metodo que foi usado num return neste fluxo.
	 * @return o metodo ou <code>null</code> se nao existir
	 */
	public Method getReturnInvocationMethod() {
		for (FieldMethodInvocation invocation : invocations) {
			if (invocation.isReturnInvocation() ) return invocation.getMethod();
		}
		return null;
	}



}
