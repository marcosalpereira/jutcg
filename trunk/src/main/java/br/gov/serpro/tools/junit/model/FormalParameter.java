package br.gov.serpro.tools.junit.model;

import java.util.List;

/**
 * Formal parameters of a method.
 */
public class FormalParameter {
    /**
     * Formal parameter name.
     */
    private String variableId;

    /**
     * Field type.
     */
    private Type type;

    /**
     * Field's annotations.
     */
    private List<Type> annotations;

    public List<Type> getAnnotations() {
        return this.annotations;
    }

    public void setAnnotations(final List<Type> annotations) {
        this.annotations = annotations;
    }

    public String getVariableId() {
        return this.variableId;
    }

    public void setVariableId(final String name) {
        this.variableId = name;
    }

    public Type getType() {
        return this.type;
    }

    public void setType(final Type type) {
        this.type = type;
    }

}
