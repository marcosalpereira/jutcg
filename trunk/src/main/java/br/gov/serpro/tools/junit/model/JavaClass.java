package br.gov.serpro.tools.junit.model;

import java.util.ArrayList;
import java.util.List;

import br.gov.serpro.tools.junit.Config;

/**
 * A class representing a Java source file.
 */
public class JavaClass {
	/**
	 * Package name.
	 */
	private String packageName;

	/**
	 * Class type.
	 */
	private Type type;

	/**
	 * Class imports.
	 */
	private List<String> imports = new ArrayList<String>();

	/**
	 * Class methods.
	 */
	private List<Method> methods = new ArrayList<Method>();

	/**
	 * Class fields.
	 */
	private List<Field> fields = new ArrayList<Field>();

	public String getPackageName() {
		return packageName;
	}

	public void setPackageName(String packageName) {
		this.packageName = packageName;
	}

	public Type getType() {
		return type;
	}

	public void setType(Type type) {
		this.type = type;
	}

	public List<String> getImports() {
		return imports;
	}

	public void setImports(List<String> imports) {
		this.imports = imports;
	}

	public List<Method> getMethods() {
		return methods;
	}

	public void setMethods(List<Method> methods) {
		this.methods = methods;
	}

	public List<Field> getFields() {
		return fields;
	}

	public void setFields(List<Field> fields) {
		this.fields = fields;
	}

	public boolean isAtView() {
		return getPackageName().startsWith(Config.getString("packagePrefix.view"));
	}

	public boolean isAtDao() {
		return getPackageName()
				.startsWith(Config.getString("packagePrefix.dao"));
	}

	public String variableNameForType() {
		if (getType() == null) return null;
		return getType().getVariableName();
	}

	public Field searchField(String fieldName) {
		for (Field f : getFields()) {
			if (f.getName().equals(fieldName)) return f;
		}
		return null;
	}

	public Method searchMethod(String methodName) {
		for (Method m : getMethods()) {
			if (m.getName().equals(methodName)) return m;
		}
		return null;
	}

}
