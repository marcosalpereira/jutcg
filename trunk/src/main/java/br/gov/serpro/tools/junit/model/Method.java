package br.gov.serpro.tools.junit.model;

import static br.gov.serpro.tools.junit.util.GeneratorHelper.lowerCaseFirstChar;
import static br.gov.serpro.tools.junit.util.GeneratorHelper.plural;
import static br.gov.serpro.tools.junit.util.GeneratorHelper.upperCaseFirstChar;

import java.util.ArrayList;
import java.util.List;

import org.apache.commons.lang.builder.EqualsBuilder;
import org.apache.commons.lang.builder.HashCodeBuilder;

import br.gov.serpro.tools.junit.util.GeneratorHelper;

/**
 * Reprsents a method in java class.
 */
public class Method {
    /**
     * Method name.
     */
    private String name;

    /**
     * Method return type.
     */
    private Type type;

    /**
     * Method's parameters.
     */
    private List<FormalParameter> formalParameters = new ArrayList<FormalParameter>();

    /**
     * Method protection.
     */
    private Protection protection;

    private List<Flow> flows = new ArrayList<Flow>();

    /**
     * Parent class.
     */
    private JavaClass javaClass;

    public List<Flow> getFlows() {
        return this.flows;
    }

    public void setFlows(final List<Flow> flows) {
        this.flows = flows;
    }

    public void setJavaClass(final JavaClass javaClass) {
        this.javaClass = javaClass;
    }

    public JavaClass getJavaClass() {
        return this.javaClass;
    }

    public String getName() {
        return this.name;
    }

    public void setName(final String name) {
        this.name = name;
    }

    public Type getType() {
        return this.type;
    }

    public void setType(final Type returnType) {
        this.type = returnType;
    }

    public List<FormalParameter> getFormalParameters() {
        return this.formalParameters;
    }

    public void setFormalParameters(final List<FormalParameter> params) {
        this.formalParameters = params;
    }

    public Protection getProtection() {
        return this.protection;
    }

    public void setProtection(final Protection protection) {
        this.protection = protection;
    }

    public boolean isVoid() {
        return this.type == null;
    }

    /**
     * <p>
     * Returns the abbreviated method signature: The method name, followed by
     * the parameter types. This is typically used in logging statements.
     * </p>
     */
    public String getLoggingSignature() {
        final StringBuilder result = new StringBuilder(getName());
        result.append('(');
        result.append(getFormalParametersAsString(", "));
        result.append(')');
        return result.toString();
    }

    /**
     * @param result
     */
    private String getFormalParametersAsString(final String separator) {
        final StringBuilder ret = new StringBuilder();
        final List<FormalParameter> params = getFormalParameters();
        for (int i = 0; i < params.size(); i++) {
            if (i > 0) {
                ret.append(separator);
            }
            ret.append(params.get(i).getType().getName());
        }
        return ret.toString();
    }

    public boolean isPrivate() {
        return this.protection == Protection.PRIVATE;
    }

    public boolean isGetter() {
    	final String fieldName;
		if (name.startsWith("is")) {
        	fieldName = GeneratorHelper.lowerCaseFirstChar(name.substring(2));

    	} else if (name.startsWith("get")) {
        	fieldName = GeneratorHelper.lowerCaseFirstChar(name.substring(3));
    	} else {
    		fieldName = null;
    	}
    	return fieldName != null && javaClass.existsField(fieldName);
    }

    public boolean isSetter() {
    	final String fieldName;
		if (name.startsWith("set")) {
        	fieldName = GeneratorHelper.lowerCaseFirstChar(name.substring(3));
    	} else {
    		fieldName = null;
    	}
    	return fieldName != null && javaClass.existsField(fieldName);
    }

    public boolean isNotVoid() {
        return !isVoid();
    }

    /**
     * The name for a method that tests this method at flow informed.
     *
     * @return the name founded
     */
    public String getNameForTest(final Flow flow) {
        String ret = "test" + upperCaseFirstChar(getName());
        if (isOverload()) {
            ret += getFormalParametersAsString("");
        }
        if (getFlows().size() > 1) {
            ret += flow.getName();
        }
        return ret;
    }

    private boolean isOverload() {
        return getJavaClass().isAnOverloadedMethod(this);
    }

    public String getVarNameFromReturningMock() {
	    if (type.isCollection()) {
	    	final Type[] generic = type.getGeneric();
	    	if (generic != null) {
	    		return lowerCaseFirstChar(plural(generic[0].getName()));
	    	}
	    	return lowerCaseFirstChar(type.getName());
	    }
	    if (type.isPrimitive()) {
	    	return lowerCaseFirstChar(this.getName());
	    }
	    return lowerCaseFirstChar(type.getName());
    }

	public String getVarNameForRealContentInvocation() {
	    if (type.isCollection()) {
	    	final Type[] generic = type.getGeneric();
	    	if (generic != null) {
	    		return lowerCaseFirstChar(plural(generic[0].getName())) + "Real";
	    	}
	    	return lowerCaseFirstChar(type.getName()) + "Real";
	    }
	    if (type.isPrimitive()) {
	    	return lowerCaseFirstChar(this.getName()) + "Real";
	    }
	    return lowerCaseFirstChar(type.getName()) + "Real";
	}




    /** {@inheritDoc} */
    @Override
    public boolean equals(final Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj instanceof Method) {
            final Method that = (Method) obj;
            return new EqualsBuilder().append(this.getLoggingSignature(),
                    that.getLoggingSignature()).isEquals();
        }

        return false;
    }

    /** {@inheritDoc} */
    @Override
    public int hashCode() {
        return new HashCodeBuilder(123, 445).append(this.getLoggingSignature()).toHashCode();
    }

    @Override
    public String toString() {
        return this.name;
    }

}
