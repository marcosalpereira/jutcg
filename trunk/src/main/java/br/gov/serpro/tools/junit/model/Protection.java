package br.gov.serpro.tools.junit.model;

/**
 * A java member visibility restriction.
 */
public enum Protection {
    DEFAULT, PRIVATE, PROTECTED, PUBLIC;

    @Override
    public String toString() {
        return this.name().toLowerCase();
    }
}
