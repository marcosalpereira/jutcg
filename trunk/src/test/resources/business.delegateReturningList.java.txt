package br.gov.esaf.sgc.business.impl;

import javax.ejb.EJB;
import javax.ejb.Stateless;

import br.gov.esaf.sgc.dao.CampoAtuacaoDao;
import br.gov.esaf.sgc.model.Pessoa;

public class PessoaBusinessBean implements PessoaBusiness {

    @EJB
    private PessoaDao pessoaDao;

    @Override
    public List<Pessoa> consultarPessoas(String parteNome) {
        return pessoaDao.selecionarPessoas(parteNome);
    }

}