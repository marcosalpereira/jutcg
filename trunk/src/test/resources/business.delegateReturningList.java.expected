package br.gov.esaf.sgc.business.impl;

import javax.ejb.EJB;
import javax.ejb.Stateless;
import br.gov.esaf.sgc.dao.CampoAtuacaoDao;
import br.gov.esaf.sgc.model.Pessoa;
import static org.easymock.EasyMock.*;
import org.junit.*;
import java.util.*;
import static org.junit.Assert.*;

/**
 * Testes unitarios para a classe {@link PessoaBusinessBean}.
 */
public class TestPessoaBusinessBean  {

/**
 * Classe sendo testada {@link PessoaBusinessBean}.
 */
private PessoaBusinessBean pessoaBusinessBean;

/**
 * Configuracoes iniciais.
 */
@Before
public void setUp() {
  pessoaBusinessBean = new PessoaBusinessBean();
}

/**
 * Cria o mock {@link PessoaDao} e seta na classe sendo testada.
 * @return o mock criado
 */
private PessoaDao criarMockPessoaDao() {
  PessoaDao mock = createStrictMock(PessoaDao.class);
  pessoaBusinessBean.setPessoaDao(mock);
  return mock;
}

/**
 * Teste para o metodo {@link PessoaBusinessBean#consultarPessoas(String)}.
 */
@Test
public void testConsultarPessoas() {

// variaveis usadas
final List<Pessoa> selecionarPessoasFromMock = Arrays.asList(new Pessoa(1));
final String parteNome = "1";

// Configurando mock
PessoaDao pessoaDao = criarMockPessoaDao();
expect(pessoaDao.selecionarPessoas(parteNome))
  .andReturn(selecionarPessoasFromMock);
replay(pessoaDao);

// invocar metodo sendo testado
final List<Pessoa> consultarPessoasReal = pessoaBusinessBean.consultarPessoas(parteNome);

// checar estados dos mocks
verify(pessoaDao);

// verificacoes do resultado do metodo sendo testado
assertEquals(selecionarPessoasFromMock, consultarPessoasReal);
}
}
